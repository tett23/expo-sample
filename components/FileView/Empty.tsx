import { Text, View } from 'native-base';
import React from 'react';

export default function Empty() {
  return (
    <View>
      <Text>empty</Text>
    </View>
  );
}
