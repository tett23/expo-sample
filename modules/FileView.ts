export const FileViewModePreview = 'preview';
export const FileViewModeEdit = 'edit';
export type FileViewModes = 'preview' | 'edit';

export type FileViewState = {
  repository: string;
  path: string;
  mode: FileViewModes;
};

function initialState(): FileViewState {
  return {
    repository: '',
    path: '',
    mode: FileViewModeEdit as typeof FileViewModeEdit,
  };
}

export default function fileView(state = initialState(), action: FileViewActions) {
  switch (action.type) {
    case SelectFileAction:
      return {
        ...state,
        ...action.payload,
      };
    case EnterEditModeAction:
      return {
        ...state,
        ...action.payload,
      };
    case EnterPreviewModeAction:
      return {
        ...state,
        ...action.payload,
      };
    default:
      return state;
  }
}

export const SelectFileAction = 'FileView/SelectFile';

export function selectFile(repository: string, path: string) {
  return {
    type: SelectFileAction as typeof SelectFileAction,
    payload: {
      repository,
      path,
    },
  };
}

export const EnterPreviewModeAction = 'FileView/EnterPreviewModeAction';

export function enterPreviewMode() {
  return {
    type: EnterPreviewModeAction as typeof EnterPreviewModeAction,
    payload: {
      mode: FileViewModePreview as typeof FileViewModePreview,
    },
  };
}

export const EnterEditModeAction = 'FileView/EnterEditModeAction';

export function enterEditMode() {
  return {
    type: EnterEditModeAction as typeof EnterEditModeAction,
    payload: {
      mode: FileViewModeEdit as typeof FileViewModeEdit,
    },
  };
}

export type FileViewActions =
  | ReturnType<typeof selectFile>
  | ReturnType<typeof enterPreviewMode>
  | ReturnType<typeof enterEditMode>;
