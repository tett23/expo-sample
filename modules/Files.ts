export const FileTypeDirectory = 'directory';
export const FileTypeFile = 'file';
export type FileTypes = 'directory' | 'file';

export type File = {
  type: FileTypes;
  repository: string;
  path: string;
  headline: string;
  tags: string[];
  variables: any[];
};

export type FilesState = {
  items: File[];
};

function initialState(): FilesState {
  return {
    items: [],
  };
}

export default function files(state = initialState(), action: FilesActions) {
  switch (action.type) {
    case FilesAddAction:
      return { ...state, items: [...state.items, action.payload] };
    case FilesRemoveAction: {
      const idx = state.items.findIndex(
        ({ repository, path }) => repository === action.payload.repository && path === action.payload.path,
      );
      if (idx === -1) {
        return state;
      }

      return {
        ...state,
        items: [...state.items.slice(0, idx), ...state.items.slice(idx + 1)],
      };
    }
    default:
      return state;
  }
}

export const FilesAddAction = 'Files/Add';

export function AddFile(file: File) {
  return {
    type: FilesAddAction as typeof FilesAddAction,
    payload: file,
  };
}

export const FilesRemoveAction = 'Files/Remove';

export function RemoveFile(repository: string, path: string) {
  return {
    type: FilesRemoveAction as typeof FilesRemoveAction,
    payload: {
      repository,
      path,
    },
  };
}

export type FilesActions = ReturnType<typeof AddFile> | ReturnType<typeof RemoveFile>;
