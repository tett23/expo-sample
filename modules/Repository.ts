export const RepositoryTypeGit = 'repository';
export const RepositoryTypeFolder = 'folder';
export type RepositoryTypes = 'repository' | 'folder';

export type Repository = {
  type: RepositoryTypes;
  name: string;
  path: string;
};
export type RepositoriesState = {
  items: Repository[];
  selectedIndex: number | null;
};

function initialState(): RepositoriesState {
  return {
    items: [],
    selectedIndex: null,
  };
}

export default function repositories(state = initialState(), action: RepositoryActions): RepositoriesState {
  switch (action.type) {
    case RepositoryAddAction:
      return { ...state, items: [...state.items, action.payload] };
    case RepositoryRemoveAction:
      return {
        ...state,
        items: [...state.items.slice(0, action.payload), ...state.items.slice(action.payload + 1)],
        selectedIndex: state.selectedIndex === action.payload ? null : state.selectedIndex,
      };
    case RepositorySelectAction:
      if (action.payload > state.items.length - 1) {
        return state;
      }

      return {
        ...state,
        selectedIndex: action.payload,
      };
    case RepositoryUnselectAction:
      return {
        ...state,
        selectedIndex: null,
      };
    default:
      return state;
  }
}

export const RepositoryAddAction = 'Repository/Add';

export function AddRepository(repo: Repository) {
  return {
    type: RepositoryAddAction as typeof RepositoryAddAction,
    payload: repo,
  };
}

export const RepositoryRemoveAction = 'Repository/RemoveRepository';

export function RemoveRepository(repoIndex: number) {
  return {
    type: RepositoryRemoveAction as typeof RepositoryRemoveAction,
    payload: repoIndex,
  };
}

export const RepositorySelectAction = 'Repository/SelectRepository';

export function SelectRepository(repoIndex: number) {
  return {
    type: RepositorySelectAction as typeof RepositorySelectAction,
    payload: repoIndex,
  };
}

export const RepositoryUnselectAction = 'Repository/UnselectRepository';

export function UnselectRepository() {
  return {
    type: RepositoryUnselectAction as typeof RepositoryUnselectAction,
    payload: {},
  };
}

export type RepositoryActions =
  | ReturnType<typeof AddRepository>
  | ReturnType<typeof RemoveRepository>
  | ReturnType<typeof SelectRepository>
  | ReturnType<typeof UnselectRepository>;
