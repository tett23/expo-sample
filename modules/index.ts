import { combineReducers, createStore } from 'redux';
import files, { FilesState, FileTypeFile } from './Files';
import fileView, { FileViewModeEdit, FileViewState } from './FileView';
import repositories, { RepositoriesState, RepositoryTypeFolder, RepositoryTypeGit } from './Repository';

export type State = {
  repositories: RepositoriesState;
  files: FilesState;
  fileView: FileViewState;
};

const reducers = combineReducers({ repositories, files, fileView });
const initialState: State = {
  repositories: {
    selectedIndex: null,
    items: [
      {
        type: RepositoryTypeGit,
        name: 'test git repo',
        path: '/',
      },
      {
        type: RepositoryTypeFolder,
        name: 'test folder',
        path: '/',
      },
    ],
  },
  files: {
    items: [
      {
        type: FileTypeFile,
        repository: 'test git repo',
        path: '/foo.md',
        headline: '',
        tags: [],
        variables: [],
      },
      {
        type: FileTypeFile,
        repository: 'test folder',
        path: '/foo.md',
        headline: '',
        tags: [],
        variables: [],
      },
    ],
  },
  fileView: {
    repository: '',
    path: '',
    mode: FileViewModeEdit as typeof FileViewModeEdit,
  },
};

export default createStore(reducers, initialState);
