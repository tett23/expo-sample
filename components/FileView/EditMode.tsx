import { Fab, Form, Icon, Textarea, View } from 'native-base';
import React, { Fragment } from 'react';
import { StyleSheet } from 'react-native';
import { File } from '../../modules/Files';

type Props = {
  file: File;
  onPressFab: () => void;
};

export default function EditMode({ file, onPressFab }: Props) {
  return (
    <Fragment>
      <View>
        <Form>
          <Textarea style={styles.textarea} rowSpan={10} placeholder="Textarea" value={file.path} />
        </Form>
      </View>
      <Fab onPress={onPressFab}>
        <Icon type="FontAwesome" name="file" />
      </Fab>
    </Fragment>
  );
}

const styles = StyleSheet.create({
  textarea: {
    height: '100%',
  },
});
