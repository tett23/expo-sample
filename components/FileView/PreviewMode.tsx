import { Fab, Icon, Text, View } from 'native-base';
import React, { Fragment } from 'react';
import { File } from '../../modules/Files';

type Props = {
  file: File;
  onPressFab: () => void;
};

export default function PreviewMode({ file, onPressFab }: Props) {
  return (
    <Fragment>
      <View>
        <Text>{file.path}</Text>
      </View>
      <Fab onPress={onPressFab}>
        <Icon type="FontAwesome" name="edit" />
      </Fab>
    </Fragment>
  );
}
