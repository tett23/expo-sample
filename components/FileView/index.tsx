import React from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { State } from '../../modules';
import { File } from '../../modules/Files';
import { enterEditMode, enterPreviewMode, FileViewModeEdit } from '../../modules/FileView';
import EditMode from './EditMode';
import Empty from './Empty';
import PreviewMode from './PreviewMode';

type Props = {
  file: File | null;
  mode: string;
  onPressEditButton: () => void;
  onPressPreviewButton: () => void;
};

export function FileView({ file, mode, onPressEditButton, onPressPreviewButton }: Props) {
  if (file == null) {
    return <Empty />;
  }

  if (mode === FileViewModeEdit) {
    return <EditMode file={file} onPressFab={onPressPreviewButton} />;
  } else {
    return <PreviewMode file={file} onPressFab={onPressEditButton} />;
  }
}

function mapStateToProps({ fileView, files }: State) {
  const file = files.items.find((item) => item.repository === fileView.repository && item.path === fileView.path);
  if (file == null) {
    return {
      file: null,
      mode: fileView.mode,
    };
  }

  return {
    file,
    mode: fileView.mode,
  };
}

function mapDispatchToProps(dispatch: Dispatch) {
  return {
    onPressEditButton() {
      dispatch(enterEditMode());
    },
    onPressPreviewButton() {
      dispatch(enterPreviewMode());
    },
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(FileView);
