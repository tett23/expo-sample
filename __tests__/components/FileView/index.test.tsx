// tslint:disable no-empty
import { shallow } from 'enzyme';
import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import { FileView } from '../../../components/FileView/index';
import { File, FileTypeFile } from '../../../modules/Files';
import { FileViewModeEdit, FileViewModePreview } from '../../../modules/FileView';

describe('', () => {
  const props = {
    file: {
      type: FileTypeFile as typeof FileTypeFile,
      repository: 'test git repo',
      path: '/foo.md',
      headline: '',
      tags: [],
      variables: [],
    },
    body: '',
    onPressEditButton: () => {},
    onPressPreviewButton: () => {},
    onChangeText: (_: File | null, __: string) => {},
  };

  describe.skip('when file is null', () => {
    it('has <Empty />', () => {
      const wrapper = shallow(<FileView {...props} file={null} mode={FileViewModeEdit} />);

      expect(wrapper.length).toBe(1);
      expect(wrapper.first().name()).toBe('Empty');
    });

    it('renders correctly', () => {
      const tree = renderer.create(<FileView {...props} file={null} mode={FileViewModeEdit} />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe.skip('when mode is FileViewModeEdit', () => {
    it('has <EditMode />', () => {
      const wrapper = shallow(<FileView {...props} mode={FileViewModeEdit} />);

      expect(wrapper.length).toBe(1);
      expect(wrapper.first().name()).toBe('EditMode');
    });

    it('renders correctly', () => {
      const tree = renderer.create(<FileView {...props} mode={FileViewModeEdit} />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe.skip('when mode is FileViewModePreview', () => {
    it('has <PreviewMode />', () => {
      const wrapper = shallow(<FileView {...props} mode={FileViewModePreview} />);

      expect(wrapper.length).toBe(1);
      expect(wrapper.first().name()).toBe('PreviewMode');
    });

    it('renders correctly', () => {
      const tree = renderer.create(<FileView {...props} mode={FileViewModePreview} />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
