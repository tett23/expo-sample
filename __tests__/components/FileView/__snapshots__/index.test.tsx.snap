// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` when file is null renders correctly 1`] = `
<View
  style={Object {}}
  virtual={undefined}
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#000",
        "fontFamily": "System",
        "fontSize": 16,
      }
    }
    uppercase={false}
    virtual={undefined}
  >
    empty
  </Text>
</View>
`;

exports[` when mode is FileViewModeEdit renders correctly 1`] = `
Array [
  <View
    style={Object {}}
    virtual={undefined}
  >
    <View
      style={Object {}}
      virtual={undefined}
    >
      <TextInput
        allowFontScaling={true}
        editable={true}
        multiline={true}
        placeholder="Textarea"
        placeholderTextColor="#575757"
        rowSpan={10}
        style={
          Object {
            "color": "#000",
            "fontSize": 15,
            "height": "100%",
            "paddingLeft": 10,
            "paddingRight": 5,
            "textAlignVertical": "top",
          }
        }
        underlineColorAndroid="rgba(0,0,0,0)"
        value="/foo.md"
        virtual={undefined}
      />
    </View>
  </View>,
  <View
    collapsable={undefined}
    style={
      Object {
        "alignItems": "center",
        "bottom": 20,
        "flexDirection": "column",
        "height": 56,
        "left": undefined,
        "position": "absolute",
        "right": 20,
        "top": undefined,
        "width": 56,
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "blue",
          "borderRadius": 28,
          "bottom": 0,
          "elevation": 4,
          "height": 56,
          "justifyContent": "center",
          "opacity": 1,
          "position": "absolute",
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.4,
          "shadowRadius": 2,
          "width": 56,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 12,
            },
            Array [
              Object {
                "color": "#000",
                "fontSize": 30,
              },
              Object {
                "color": "#000",
                "fontSize": 30,
              },
              Object {
                "color": "#fff",
                "fontSize": 24,
                "lineHeight": 27,
              },
            ],
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
        type="FontAwesome"
        virtual={undefined}
      >
        
      </Text>
    </View>
  </View>,
]
`;

exports[` when mode is FileViewModePreview renders correctly 1`] = `
Array [
  <View
    style={Object {}}
    virtual={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#000",
          "fontFamily": "System",
          "fontSize": 16,
        }
      }
      uppercase={false}
      virtual={undefined}
    >
      /foo.md
    </Text>
  </View>,
  <View
    collapsable={undefined}
    style={
      Object {
        "alignItems": "center",
        "bottom": 20,
        "flexDirection": "column",
        "height": 56,
        "left": undefined,
        "position": "absolute",
        "right": 20,
        "top": undefined,
        "width": 56,
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "blue",
          "borderRadius": 28,
          "bottom": 0,
          "elevation": 4,
          "height": 56,
          "justifyContent": "center",
          "opacity": 1,
          "position": "absolute",
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.4,
          "shadowRadius": 2,
          "width": 56,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 12,
            },
            Array [
              Object {
                "color": "#000",
                "fontSize": 30,
              },
              Object {
                "color": "#000",
                "fontSize": 30,
              },
              Object {
                "color": "#fff",
                "fontSize": 24,
                "lineHeight": 27,
              },
            ],
            Object {
              "fontFamily": "FontAwesome",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
        type="FontAwesome"
        virtual={undefined}
      >
        
      </Text>
    </View>
  </View>,
]
`;
